app/src/androidTest/java/ch/epfl/balelecbud/util/cache/FilesystemCacheTest.java

```diff
@@ -0,0 +1,159 @@
+import ch.epfl.balelecbud.util.database.Database;
+import ch.epfl.balelecbud.util.database.MockDatabase;
+import ch.epfl.balelecbud.util.database.MyGeoClause;
+import ch.epfl.balelecbud.util.database.MyQuery;
+import ch.epfl.balelecbud.util.database.MyWhereClause;
+
+import static ch.epfl.balelecbud.util.database.Database.DOCUMENT_ID_OPERAND;
+import static ch.epfl.balelecbud.util.database.MockDatabase.assertQueryMapResults;
+import static ch.epfl.balelecbud.util.database.MockDatabase.assertQueryResults;
+import static ch.epfl.balelecbud.util.database.MockDatabase.axel;
+import static ch.epfl.balelecbud.util.database.MockDatabase.karim;
+import static ch.epfl.balelecbud.util.database.MockDatabase.slot1;
+import static ch.epfl.balelecbud.util.database.MockDatabase.slot2;
+import static ch.epfl.balelecbud.util.database.MockDatabase.token1;
+import static ch.epfl.balelecbud.util.database.MyWhereClause.Operator.EQUAL;
+
+@RunWith(AndroidJUnit4.class)
+public class FilesystemCacheTest {
+
+    private Cache appCache = BalelecbudApplication.getAppCache();
+
+    @Before
+    public void setup() {
+        BalelecbudApplication.setRemoteDatabase(MockDatabase.getInstance());
+        appCache.flush();
+    }
+
+    @Test
+    public void containsCorrectForCollectionRetrieval() throws IOException {
+        MyQuery query = new MyQuery(Database.POINT_OF_INTEREST_PATH, Collections.emptyList());
+        Assert.assertFalse(appCache.contains(query));
+        appCache.put(Database.CONCERT_SLOTS_PATH, "0", slot1);
+        appCache.put(Database.CONCERT_SLOTS_PATH, "1", slot2);
+        query = new MyQuery(Database.CONCERT_SLOTS_PATH, Collections.emptyList());
+        Assert.assertTrue(appCache.contains(query));
+    }
+
+    @Test
+    public void containsCorrectForPerIdRetrieval() throws IOException {
+        appCache.put(Database.USERS_PATH, karim.getUid(), karim);
+        MyQuery query = new MyQuery(Database.USERS_PATH, new MyWhereClause(DOCUMENT_ID_OPERAND, EQUAL, karim.getUid()));
+        Assert.assertTrue(appCache.contains(query));
+        query = new MyQuery(Database.USERS_PATH, new MyWhereClause(DOCUMENT_ID_OPERAND, EQUAL, axel.getUid()));
+        Assert.assertFalse(appCache.contains(query));
+    }
+
+    @Test
+    public void containsWithGeoclauseQueryFalse() throws IOException {
+        appCache.put(Database.LOCATIONS_PATH, "0", Location.DEFAULT_LOCATION);
+        MyQuery query = new MyQuery(Database.LOCATIONS_PATH, new MyGeoClause(0, 0, 1));
+        Assert.assertFalse(appCache.contains(query));
+    }
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void containsWithFieldWhereClauseFalse() throws IOException {
+        appCache.put(Database.USERS_PATH, karim.getUid(), karim);
+        MyQuery query = new MyQuery(Database.USERS_PATH, new MyWhereClause("email", EQUAL, karim.getEmail()));
+        appCache.contains(query);
+    }
+
+    @Test
+    public void putGetWorksCorrectlyWithCustomTypes() throws Throwable {
```

Comment:
Also avoid using "WorksCorrectly" in test method names, you can rename to "testPutAndGetWithCustomTypes". Same for line 102